doc/readmsg.1
.if n \{\
.	ds ct "
.\}
.if t \{\
.	ds ct \\(co
.\}
.TH READMSG 1L "Elm Version 2.5" "USENET Community Trust"
.SH NAME
readmsg - extract messages from a mail folder
.SH SYNOPSIS
.B readmsg
[\fB\-anhp\fR]
[\fB\-f\fR folder]
[\fB\-w\fR weedlist]
[selection ...]
.SH DESCRIPTION
.I Readmsg
extracts selected mail messages from a mail folder.  One helpful use of
the program is when you are composing a response to a mail message in an
external editor.  In this case, you can run
.I readmsg
to pull a copy of the original message into the editing buffer.
.P
When you run
.I readmsg
from within
.I elm
(e.g. from a subshell escape or in an external editor while composing
a mail reply) it behaves a bit differently from when you run it directly
(e.g. from a shell command line).  We will first describe its normal
behavior, and then describe how it is different when you run it under
.IR elm .
.P
You tell
.I readmsg
which messages to extract with the
.I selection
argument.  There are a couple of possible different ways to specify
the
.IR selection .
.IP "1."
A lone ``*'' means select all messages in the mailbox.
.IP "2."
A list of message numbers may be specified.  Values of ``0'' and
``$'' in the list both mean the last message in the mailbox.  For
example:
.ft CW
.sp
	readmsg 1 3 0
.sp
.ft P
extracts three messages from the folder:  the first, the third, and
the last.
.IP "3."
Finally, the
.I selection
may be some text to match.  This will select a mail message which
.I exactly
matches the specified text.  For example,
.ft CW
.sp
	readmsg staff meeting
.sp
.ft P
extracts the message which contains the words ``staff meeting.''  Note
that it will
.I not
match a message containing ``Staff Meeting'' \(en the matching is case
sensitive.  Normally only the first message which matches the pattern
will be printed.  The
.B \-a
option discussed in a moment changes this.
.P
The
.B \-f
flag indicates that you'd rather use the folder specified
rather than the default incoming mailbox.  The specified
.I folder
can be a filename or a specification such as ``=sentmail''.
.P
The
.B \-w
flag tailors the list of mail headers that are weeded and displayed.
Header weeding is discussed below.
.P
The
.B \-h
flag instructs the program to ignore the header weeding list, and
include the entire header of the matched message or messages when
displaying their text.
.P
The
.B \-n
flag instructs the program to ignore the header weeding list, and
exclude \fIall\fR headers.  This is used mostly for extracting files
mailed and such.
.P
The
.B \-p
flag indicates that the program should
put form-feeds (control-L) between message headers.
.P
The
.B \-a
flag indicates that all messages which match the pattern
specified on the command line should be printed, not just the first.
If a pattern was not specified on the command line then this flag has
no effect.
.P
When you run
.I readmsg
under
.I elm
(say in the context of an external editor) the behavior
will be different from that described above as follows.
.IP "1."
The default mail folder will be the folder you are currently
examining in
.I elm
and not necessarily your incoming mail folder.
.IP "2."
You do not need to specify a
.I selection
on the command line.  If you omit the
.I selection
then
.I readmsg
will extract the message(s) you have selected in Elm.  If you have
tagged any messages then this would be all of the tagged messages,
otherwise it would be the message you are currently examining.
.IP "3."
Normally the message numbers
.I readmsg
uses are in mailbox order.  When you call
.I readmsg
under
.I elm
and do not override the folder selection with the
.B \-f
option, then message numbers will be sorted as they are displayed on
the
.I elm
message index screen.
.P
Normally,
.I readmsg
selects and displays certain headers from the message.  By default,
the list of selected headers includes:
.P
.RS
.nf
\(bu The ``Subject:'' header.
\(bu The ``From:'' header.
\(bu The ``To:'' header.
\(bu The ``Cc:'' header.
\(bu The ``Date:'' header.
\(bu Any header starting with ``Apparently-''.
.fi
.RE
.P
The header weeding may be changed with the
.B \-w
option.  The argument to this option is a list of header names,
separated by whitespace or commas.
.I readmsg
compares each header to the ones in this list, and if a match is found
the header is displayed.  The list entries can contain partial header
names.  For instance, an entry of ``From'' would match both the
``From\s-1<SPACE>\s0'' line at the top of the message, as well as any
``From:'' header.  Matching is case insensitive.  An underscore (``_'')
may be used to represent a space, so specifying ``From_'' will select
the ``From\s-1<SPACE>\s0'' line but not the ``From:'' header.  A list
entry may be preceded by an exclamation point to suppress the header.
Thus, a specification of ``!From_\0From'' means about the same thing
as ``From:''.  When no header weeding options are specified, the
default action corresponds to:
.P
.ft CW
	readmsg -w "Subject: From: To: Cc: Apparently- Date:"
.ft 1
.SH "EXAMPLES"
First off, to use this from within \fBvi\fR to include the text of the 
current message, you could use the command:
.nf

	:r !readmsg

.fi
(as you hit the ':' the editor will put you at the bottom of the screen
with the ':' prompt).  The space following ':r' is required.
.sp 2
Let's look at something more interesting, however;
.sp
Suppose you have the mail file;
.nf

   From joe Jun 3 1986 4:45:30 MST
   Subject: hello
   
   Hey Guy!  Wanta go out and have a milk this evening?
   
   Joe
   
   From john Jun 3 1986 4:48:20 MST
   Subject: Dinner at Eight
   From: John Dinley <xyz!john>

   Remember you should show up about eight, okay?

		   - John D -

   From xxzyz!cron Jun 3 1986 5:02:43 MST

   Cannot connect to server: blob
   Job 43243 deleted from queue.
   
.fi
The following commands will result in;
.nf

  $ readmsg 2		
  [ display the second message, from John ]

  $ readmsg 		
  [ an error, unless we're calling from \fIelm\fR ]

  $ readmsg BLOB	
  [ no match - case sensitive! ]

  $ readmsg -h connect to server 
  [ displays third message, including headers ]

.fi
.SH FILES
/usr/mail/<username>   		The incoming mail
.br
$ELMSTATE			Status information from \fIelm\fR
.SH AUTHOR
Elm Development Group
.SH SEE ALSO
newmail(1L), elm(1L)
.SH BUGS
The '*' metacharacter doesn't always work as expected!
.br
Perhaps the pattern matching should be case insensitive?
.br
It might be confusing that messages are sorted when running under
.I elm
with the current folder, and in mailbox order for all other cases.
When
.I readmsg
is run standalone, messages may be truncated at lines beginning with
"From".  This is not a problem when
.I readmsg
is spawned under
.IR elm
because the status information file created by
.I elm
says exactly how long each message is.
.SH BUG REPORTS TO
Bill Pemberton  flash@virginia.edu
.SH COPYRIGHTS
\fB\*(ct\fRCopyright 1988-1995 by The USENET Community Trust
.br
Derived from Elm 2.0, \fB\*(ct\fR Copyright 1986, 1987 by Dave Taylor


Changes
Changes in 2.5 PL8

- fixed a hang when hitting certain 8-bit chars in the From line

- fixed a file descriptor leak (from Andreas Luik)

- removed all Expires: processing


Changes in 2.5 PL7

- a few display issues fixed (from Kenneth Reek)

- buffer overflow in frm fixed

Changes in 2.5 PL6

- Fixed message-id in the In-Reply-To header.

- Actually show the -r option in the usage

- Whacked a few more places where we might follow a symlink in tmpdir

- Fastmail didn't grok commas correctly.

- Change lock.c to not use errno unless there's an errno

Changes in 2.5 PL5

- A few more buffer overruns fixed.

Changes in 2.5 PL4

- little bug in aliasdb.c (from Scot W. Hetzel <scot@genroco.com>)

- Message-Id: could overflow (from Darin Wayrynen <darin@deru.net>)

- Fixed the permissions setting problem with the edit mailbox function

- Added a ResizeScreen() call after show_msg() calls to hopefully deal wit
  SIGWINCH problems (inspired from Yuke Zhuge <zhuge@stone.pdx.informix.com>)
  
- Fixed problems under SunOS 4.1.3 (from Paul J. Menchini <mench@mench.com>)

- Fixed content encoding (from David Woolley  <david@djwhome.demon.co.uk>)

- Attempt to stop crashes with long recipient list (from Gilles Detillieux 
  <grdetil@scrc.umanitoba.ca>)
  

Changes in 2.5 PL3

- Fixed the declaration of outbound_from_name (from Richard Jackson 
  <rjackson@gmu.edu>)

- Fixed the leapyear macro (from Gerhard Niklasch)
  
- Fixed editing of a long CC line (from Derek Godfrey <djg@sjc.ascend.com>)

- Mail that lands in the sentmail with attachments now has the 
  Content-disposition header text added to the message (from Dave 
  Spensley)
  
- Made allto a bit longer (suggested by Dave Spensley 
  <davesp@typhoon.xnet.com>)

- Fixed Configure to find (again) fcntl on Linux

- can_access will now allow you to write to a char device (/dev/null)

Changes in 2.5 PL2

- Fixed (again, maybe?) savebyalias from Adam Benjamin <aebenjam@opentext.com>

- Fixed bug were an alias delete would crash elm if the mailbox was empty

Changes in 2.5 PL1

- Fixed savebyalias from Michael D Hughes <mdh@logcabin.woods.bridge.com> 
  and Robert Schwartz <rjs@elt.com>
  
- Changes for 8 bit clean from H.J. Lu <hjl@varesearch.com>

Changes in 2.5 PL0

- One minor change to make an easier compile on SparcLinux.

Changes in 2.5 PL0pre8

- Y2K stuff from Ed Symanzik <zik@msu.edu>

Changes in 2.5 PL0pre7

- minor change to deal with egcs

Changes in 2.5 PL0pre6

- newalias nows does a geteuid() instead of getuid().  This is to lets 
  SysAdmins do "su <user> -c 'newalias'"
  

Changes in 2.5 PL0pre5

- Fix for GECOS entries with comma-delimited sub-fields from Chip 
  Rosenthal <chip@unicom.com>.

- Fix? for keys getting screw up after a ^Z from Chip Rosenthal
  <chip@unicom.com>.


Changes in 2.5 PL0pre4

- A few changes for BSD from Chip Rosenthal <chip@unicom.com>.

- A few changes to Configure suggested by Chip Rosenthal <chip@unicom.com>.


Changes in 2.5 PL0pre3

- Configure didn't see fcntl locking on Linux.  Fix from Sean McDermott 
  <spm@intel.com>


Changes in 2.5 beta 3

- Fixed a problem where elm would try to close a file twice when
  sending an attachment (from reichler@ews.uiuc.edu)

- Added docs for environment variables to the man page (from Eric
  Haszlakiewicz <nimenees@uaccess.net>)

Changes in 2.5 beta 2

- IncomingFolders no longer set Magic Mode on for everything

- Configure should now deal with Linux.

- Make //pattern honor current limit (from Keith Neufeld<neufeld@pvi.org>)

- year 2000 fix.....?

- Building on Solaris requires /usr/ccs/bin to be in the PATH

Changes in 2.5 beta 1

- b)ounce honor DONT_ADD_FROM (from Keith Neufeld<neufeld@pvi.org>)

- removed [encode] display and reply (from Keith Neufeld<neufeld@pvi.org>)

- make right/left arrows work in the file browser the same as in the
  message index. (from Keith Neufeld<neufeld@pvi.org>)

- fixed a bug in the incomingfolders code (it wouldn't save) (from Bob Schwartz
  <rjs@elt.com>)
  
- fixed the meaning of noheader for forwards (from Keith Neufeld<neufeld@pvi.org>)



Changes in 2.5 alpha 11

- fixed a bug in the alias code (thanks to Keith Neufeld<neufeld@pvi.org>)

- fixed the generation of temp file names in newmbox.c

- allowed longer Cc: lines.  This was due to a crash with a group reply.

- added incomingfolders to the elmrc.  This is a list of folders that will
  automatically be opened with "magic mode" on. (inspired by a patch from
  Cristophe Kalt <kalt@ensta.fr>)


Changes in 2.5 alpha 10

- Press u on a message that is read but not deleted will mark the message as
  unread (from Larry Schwimmer <rosebud@cyclone.Stanford.EDU>).

- mmencode.1 man page is now installed with others (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- Magic mode stuff is now in help and NLS (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- allow elimination of aliases from larger aliases (To: bigalias -subalias).
  Based on incomplete patch by Waldi Ravens <waldi@moacs.indiv.nl.net>.  Also
  removes redundant declarations of get_alias_address(), which is declared in
  elm_globals.h (from Keith Neufeld<neufeld@pvi.org>)

- On an entry field that normally clears to be retyped from scratch as soon

utils/readmsg.c


/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 1.7 $   $State: Exp $
 *
 *                      Copyright (c) 1988-1995 USENET Community Trust
 *			Copyright (c) 1986,1987 Dave Taylor
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *      Bill Pemberton, Elm Coordinator
 *      flash@virginia.edu
 *
 *******************************************************************************
 * $Log: readmsg.c,v $
 * Revision 1.7  1997/10/20  20:24:37  wfp5p
 * Incomingfolders no longer set Magic mode on for all remaining folders.
 *
 * Revision 1.6  1996/03/14  17:30:11  wfp5p
 * Alpha 9
 *
 * Revision 1.5  1995/09/29  17:42:49  wfp5p
 * Alpha 8 (Chip's big changes)
 *
 * Revision 1.4  1995/09/11  15:19:36  wfp5p
 * Alpha 7
 *
 * Revision 1.3  1995/06/01  13:13:33  wfp5p
 * Readmsg was fixed to work correctly if called from within elm.  From Chip
 * Rosenthal <chip@unicom.com>
 *
 * Revision 1.2  1995/04/20  21:02:10  wfp5p
 * Added the showreply feature and emacs key bindings.
 *
 * Revision 1.1.1.1  1995/04/19  20:38:39  wfp5p
 * Initial import of elm 2.4 PL0 as base for elm 2.5.
 *
 ******************************************************************************/

/** This program extracts messages from a mail folder.  It is particularly
    useful when the user is composting a mail response in an external
    editor.  He or she can call this program to pull a copy of the original
    message (or any other message in the folder) into the editor buffer.

    One of the first things we do is look for a folder state file.
    If we are running as a subprocess to Elm this file should tell us
    what folder is currently being read, the seek offsets to all the
    messages in the folder, and what message(s) in the folder is/are
    selected.  We will load in all that info and use it for defaults,
    as applicable.

    If a state file does not exist, then the default is to show whatever
    messages the user specifies on the command line.  Unless specified
    otherwise, this would be from the user's incoming mail folder.

    Even if a state file exists, the user can override the defaults
    and select a different set of messages to extract, or select an
    entirely different folder.

    Messages can be selected on the command line as follows:

	readmsg [-options] *

	    Selects all messages in the folder.

	readmsg [-options] pattern ...

	    Selects messsage(s) which match "pattern ...".  The selected
	    message will contain the "pattern ..." somewhere within
	    the header or body, and it must be an exact (case sensitive)
	    match.  Normally selects only the first match.  The "-a"
	    selects all matches.

	readmsg [-options] sel ...

	    where:  sel == a number -- selects that message number
		    sel == $ -- selects last message in folder
		    sel == 0 -- selects last message in folder

    The undocumented "-I" option is a kludge to deal with an Elm race
    condition.  The problem is that Elm does piping/printing/etc. by
    running "readmsg|command" and placing the mail message selection
    into a folder state file.  However, if the "command" portion of
    the pipeline craps out, Elm might regain control before "readmsg"
    completes.  The first thing Elm does is unlink the folder state
    file.  Thus "readmsg" can't figure out what to do -- there is no
    state file or command line args to select a message.  In this
    case, "readmsg" normally gives a usage diagnostic message.  The
    "-I" option says to ignore this condition and silently terminate.

**/

#define INTERN
#include "elm_defs.h"
#include "s_readmsg.h"

/* level of headers to display */
#define HDRSEL_ALL	1
#define HDRSEL_WEED	2
#define HDRSEL_NONE	3

#define metachar(c)	(c == '=' || c == '+' || c == '%')

/* increment for growing dynamic lists */
#define ALLOC_INCR	256

/* program name for diagnostics */
char *prog;

/*
 * The "folder_idx_list" is a list of seek offsets into the folder,
 * indexed by message number.  The "folder_size" value indicates the
 * number of messages in the folder, as well as the length of the index.
 * This index will be used if we need to access messages by message
 * number.  If possible, the index will be loaded from the external state
 * file, otherwise we will have to make a pass through the entire folder
 * to build up the index.  The index is not needed if we are printing
 * everything ("*" specified on the command line) or if we are searching
 * for pattern match.
 *
 * The "folder_clen_list" contains the content lengths for the messages.
 * This information is available *only* if we were spawned from Elm and
 * the content lengths have been retrieved from the folder state file.
 * If this information is available, it will prevent truncated messages
 * due to embedded "From" lines.  If content length information is not
 * available, this list will be left NULL.
 */
int folder_size;
long *folder_idx_list;
long *folder_clen_list;

/*
 * Information on a folder message.
 */
struct mssg_info {
	long folder_pos;	/* start of message seek position	*/
	long content_length;	/* message body length, or 0 if unknown	*/
};


/*
 * Header weeding information.
 */

#define DFLT_WEED_LIST	"Subject: From: To: Cc: Apparently- Date:"

struct weed_header_info {
    char *hdrname;	/* string to compare against header	*/
    int hdrlen;		/* characters in "hdrname"		*/
    int hdrskip;	/* skip (instead of accept) on match?	*/
    struct weed_header_info *next;
};

/* head of linked list of weeded headers */
struct weed_header_info Weedlist;

/* is default to accept headers NOT in weed list? */
int weed_dflt_accept = FALSE;


/*
 * local procedures
 */
void usage_error P_((void));
void malloc_fail_handler P_((const char *, unsigned));
void load_folder_index P_((FILE *));
int print_patmatch_mssg P_((FILE *, const char *, int, int, int));
int print_mssg P_((FILE *, const struct mssg_info *, int, int));
void weed_headers P_((const char *, int, FILE *));
void setup_weed_info P_((const char *));
int get_mssg_info P_((struct mssg_info *, int));


extern char *optarg;			/* for parsing the ...		*/
extern int   optind;			/*  .. starting arguments	*/


void usage_error()
{
    fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet, ReadmsgUsage,
	"Usage: %s [-anhp] [-f Folder] [-w weedlist] {MessageNum ... | pattern | *}\n"),
	prog);
    exit(1);
}


main(argc, argv)
int argc;
char *argv[];
{
    struct folder_state fstate;	/* information from external state file	*/
    char folder_name[SLEN];	/* pathname to the mail folder		*/
    int fstate_valid;		/* does "fstate" have valid info?	*/
    int hdr_disp_level;		/* amount of headers to show		*/
    int do_page_breaks;		/* true to formfeed between messages	*/
    int do_all_matches;		/* true to show all mssgs which match pat*/
    int ign_no_request;		/* terminate if no actions requested	*/
    int exit_status;		/* normally zero, set to one on error	*/
    FILE *fp;			/* file stream for opened folder	*/
    struct mssg_info minfo;	/* information on message to extract	*/
    char *sel_weed_str;		/* header weeding specification		*/
    char buf[SLEN];
    int i;

    initialize_common();

    prog = argv[0];
    folder_name[0] = '\0';	/* no folder specified yet		*/
    folder_size = -1;		/* message index not loaded yet		*/
    sel_weed_str = DFLT_WEED_LIST; /* default header weeding spec	*/
    hdr_disp_level = HDRSEL_WEED; /* only display interesting headers	*/
    do_page_breaks = FALSE;	/* suppress formfeed between mssgs	*/
    do_all_matches = FALSE;	/* only show 1st mssg which matches pat	*/
    ign_no_request = FALSE;	/* no action requested is an error	*/
    exit_status = 0;		/* will set nonzero on error		*/

    /* see if an external folder state file exists */
    if (load_folder_state_file(&fstate) != 0) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
	    ReadmsgStateFileCorrupt,
	    "%s: Elm folder state file appears to be corrupt!\n"), prog);
	exit(1);
    }
    fstate_valid = (fstate.folder_name != NULL);

    /* crack the command line */
    while ((i = getopt(argc, argv, "af:hnpw:I")) != EOF) {
	switch (i) {
	case 'a' :
	    do_all_matches = TRUE;
	    break;
	case 'f' :
	    strcpy(folder_name, optarg);
	    if (metachar(folder_name[0]) && !expand(folder_name)) {
		fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
		    ReadmsgCannotExpandFolderName,
		    "%s: Cannot expand folder name \"%s\".\n"),
		    prog, folder_name);
		exit(1);
	    }
	    folder_size = -1;	/* zot out index info from extern state file */
	    break;
	case 'h' :
	    hdr_disp_level = HDRSEL_ALL;
	    break;
	case 'n' :
	    hdr_disp_level = HDRSEL_NONE;
	    break;
	case 'p' :
	    do_page_breaks = TRUE;
	    break;
	case 'w':
	    hdr_disp_level = HDRSEL_WEED;
	    sel_weed_str = optarg;
	    break;
	case 'I':
	    ign_no_request = TRUE;
	    break;
	default:
	    usage_error();
	}
    }

    /* figure out where the folder is */
    if (folder_name[0] == '\0') {
	if (fstate_valid)
	    strcpy(folder_name, fstate.folder_name);
	else
	    strcpy(folder_name, incoming_folder);
    }

    /* open up the message folder */
    if ((fp = fopen(folder_name, "r")) == NULL) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
	    ReadmsgFolderEmpty, "%s: Folder \"%s\" is empty.\n"),
	    prog, folder_name);
	exit(0);
    }

    /* if this is the folder in the state file then grab the index */
    if (fstate_valid && strcmp(folder_name, fstate.folder_name) == 0) {
	folder_size = fstate.num_mssgs;
	folder_idx_list = fstate.idx_list;
	folder_clen_list = fstate.clen_list;
    } else {
	fstate_valid = FALSE;
    }

    /* process the header weeding specification */
    if (hdr_disp_level == HDRSEL_WEED)
	setup_weed_info(sel_weed_str);

    /* if no selections on cmd line then show selected mssgs from state file */
    if (argc == optind) {
	if (!fstate_valid) {
		/* no applicable state file or it didn't select anything */
		if (ign_no_request)
		    exit(0);
		usage_error();
	}
	if (folder_size < 1) {
	    fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet, ReadmsgFolderEmpty,
		"%s: Folder \"%s\" is empty.\n"), prog, folder_name);
	    exit(0);
	}
	for (i = 0 ; i < fstate.num_sel ; ++i) {
	    if (get_mssg_info(&minfo, fstate.sel_list[i]) != 0)
		exit_status = 1;
	    else if (print_mssg(fp, &minfo, hdr_disp_level, do_page_breaks)!=0)
		exit_status = 1;
	}
	exit(exit_status);
    }

    /* see if we are trying to match a pattern */
    if (index("0123456789$*", argv[optind][0]) == NULL) {
	strcpy(buf, argv[optind]);
	while (++optind < argc)
		strcat(strcat(buf, " "), argv[optind]);
	if (print_patmatch_mssg(fp, buf, do_all_matches,
			hdr_disp_level, do_page_breaks) != 0)
	    exit_status = 1;
	exit(exit_status);
    }

    /* if we do not have an index from the state file then go build one */
    if (folder_size < 0)
	load_folder_index(fp);

    /* make sure there is something there to look at */
    if (folder_size < 1) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet, ReadmsgFolderEmpty,
	    "%s: Folder \"%s\" is empty.\n"), prog, folder_name);
	exit(0);
    }

    /* see if all messages should be shown */
    if (argc-optind == 1 && strcmp(argv[optind], "*") == 0) {
	for (i = 1 ; i <= folder_size ; ++i) {
  	    if (get_mssg_info(&minfo, i) != 0)
		exit_status = 1;
	    else if (print_mssg(fp, &minfo, hdr_disp_level, do_page_breaks)!=0)
		exit_status = 1;
	}
	exit(exit_status);
    }

    /* print out all the messages specified on the command line */
    for ( ; optind < argc ; ++optind) {

	/* get the message number */
	if (strcmp(argv[optind], "$") == 0 || strcmp(argv[optind], "0") == 0) {
	    i = folder_size;
	} else if ((i = atoi(argv[optind])) == 0) {
	    fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
		ReadmsgIDontUnderstand,
		"%s: I don't understand what \"%s\" means.\n"),
		prog, argv[optind]);
	    exit(1);
	}

	/*print it */
	if (get_mssg_info(&minfo, i) != 0)
	    exit_status = 1;
	else if (print_mssg(fp, &minfo, hdr_disp_level, do_page_breaks)!=0)
	    exit_status = 1;
    }

    exit(exit_status);
    /*NOTREACHED*/
}


/*
 * Scan through the entire folder and build an index of seek offsets.
 */
void load_folder_index(fp)
FILE *fp;
{
    long offset;
    int alloc_size;
    char buf[SLEN];
    int buf_len, last_buf_len, was_empty_line;

    /* zero out the folder seek offsets index */
    folder_size = 0;
    alloc_size = 0;
    folder_idx_list = NULL;
    /* pretend there was an empty line before the first line of the mailbox */
    last_buf_len = 1;

    /* we are being lazy - we don't bother figuring out content lengths */
    folder_clen_list = NULL;

    /* go through the folder a line at a time looking for messages */
    rewind(fp);
    while (offset = ftell(fp), (buf_len = mail_gets(buf, sizeof(buf), fp)) != 0) {
	was_empty_line = last_buf_len == 1;
	last_buf_len = buf_len;

	if (was_empty_line && strbegConst(buf, "From ") &&
	    real_from(buf, (struct header_rec *)NULL))
	{
	    if (folder_size >= alloc_size) {
		alloc_size += ALLOC_INCR;
		folder_idx_list = (long *) safe_realloc((char *)folder_idx_list,
		    alloc_size*sizeof(long));
	    }
	    folder_idx_list[folder_size++] = offset;
	}
    }

}


/*
 * Scan through a folder and print message(s) which match the pattern.
 */
int print_patmatch_mssg(fp, pat, do_all_matches, hdr_disp_level, do_page_breaks)
FILE *fp;
const char *pat;
int do_all_matches;
int hdr_disp_level;
int do_page_breaks;
{
    long offset, mssg_idx;
    struct mssg_info minfo;
    char buf[VERY_LONG_STRING];
    int look_for_pat;
    int buf_len, last_buf_len, was_empty_line;

    /*
     * This flag ensures that we don't reprint a message if a single
     * message matches the pattern several times.  We turn it on when
     * we get to the top of a message.
     */
    look_for_pat = FALSE;
    /* pretend there was an empty line before the first line of the mailbox */
    last_buf_len = 1;

    rewind(fp);
    while (offset = ftell(fp), (last_buf_len = mail_gets(buf, sizeof(buf), fp)) != 0) {

	was_empty_line = last_buf_len == 1;
	last_buf_len = buf_len;
	if (was_empty_line && strbegConst(buf, "From ") &&
	    real_from(buf, (struct header_rec *)NULL))
	{
	    mssg_idx = offset;
	    look_for_pat = TRUE;
	}

	if (look_for_pat && strstr(buf, pat) != NULL) {
	    minfo.folder_pos = ftell(fp);
	    minfo.content_length = 0L;		/* we don't know this */
	    if (print_mssg(fp, &minfo, hdr_disp_level, do_page_breaks) != 0)
		return -1;
	    fseek(fp, minfo.folder_pos, 0);
	    if (!do_all_matches)
		break;
	    look_for_pat = FALSE;
	}

    }

    return 0;
}


/*
 * Print the message at the indicated location.
 */
int print_mssg(fp, minfo, hdr_disp_level, do_page_breaks)
FILE *fp;
const struct mssg_info *minfo;
int hdr_disp_level;
int do_page_breaks;
{
    char buf[SLEN];
    int first_line, is_seperator, in_header, buf_len, last_buf_len, newlines, was_empty_line;
    long curr_offset, stop_offset;
    static int num_mssgs_listed = 0;

    in_header = TRUE;
    first_line = TRUE;
    stop_offset = 0L;

    if (num_mssgs_listed++ == 0)
	; /* no seperator before first message */
    else if (do_page_breaks)
	putchar('\f');
    else
	puts("\n------------------------------------------------------------------------------\n");

    /* move to the beginning of the selected message */
    if (fseek(fp, minfo->folder_pos, 0) != 0) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet, ReadmsgCannotSeek,
	    "%s: Cannot seek to selected message. [offset=%ld]\n"),
	    prog, minfo->folder_pos);
	return -1;
    }

    /* we will chop off newlines at the end of the message */
    newlines = 0;
    /* pretend there was an empty line before the first line of the mailbox */
    last_buf_len = 1;

    /* print the message a line at a time */
    while (curr_offset = ftell(fp), (buf_len = mail_gets(buf, SLEN, fp)) != 0) {

	was_empty_line = last_buf_len == 1;
	last_buf_len = buf_len;

	if (stop_offset == 0L || curr_offset >= stop_offset) {
	    is_seperator = was_empty_line && strbegConst(buf, "From ") &&
		       real_from(buf, (struct header_rec *)NULL);

	} else {
	    is_seperator = FALSE;
	}

	/* the first line of the message better be the seperator */
	/* next time we encounter the seperator marks the end of the message */
	if (first_line) {
	    if (!is_seperator) {
		fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
		    ReadmsgCannotFindStart,
		    "%s: Cannot find start of selected message. [offset=%ld]\n"),
		    prog, minfo->folder_pos);
		return -1;
	    }
	    first_line = FALSE;
	} else {
	    if (is_seperator)
		break;
	}

	/* just accumulate newlines */
	if (buf[0] == '\n' && buf[1] == '\0') {
	    if (in_header) {
		if (minfo->content_length > 0)
		    stop_offset = curr_offset + minfo->content_length;
		in_header = FALSE;
	    }
	    ++newlines;
	    continue;
	}

	if (in_header) {
	    switch (hdr_disp_level) {
	    case HDRSEL_NONE:
		break;
	    case HDRSEL_WEED:
		weed_headers(buf, buf_len, stdout);
		break;
	    case HDRSEL_ALL:
	    default:
		fwrite(buf, 1, buf_len, stdout);
		break;
	    }
	} else {
	    while (--newlines >= 0)
		putchar('\n');
	    newlines = 0;
	    fwrite(buf, 1, buf_len, stdout);
	}

    }

    /* make sure we didn't do something like seek beyond the */
    /* end of the folder and thus never get anything to print */
    if (first_line) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet, ReadmsgCannotFindStart,
	    "%s: Cannot find start of selected message. [offset=%ld]\n"),
	    prog, minfo->folder_pos);
	return -1;
    }

    return 0;
}


/*
 * Weed out and display selected headers.
 */
void weed_headers(buf, buflen, fp)
const char *buf;
int buflen;
FILE *fp;
{
    struct weed_header_info *w;
    static int accept_header = FALSE;

    /* if this is a new header, see if we want to accept it */
    if (!isspace(buf[0])) {
	accept_header = weed_dflt_accept;   
	for (w = Weedlist.next ; w != NULL ; w = w->next) {  
	    if (strincmp(buf, w->hdrname, w->hdrlen) == 0) {
		accept_header = !w->hdrskip;
		break;
	    }
	}
    }

    /* output accepted headers */
    if (accept_header)
	fwrite(buf, 1, buflen, fp);
}


/* 
 * Initialize the header weeding info.
 *
 * The header weeding info is a linked list of (struct weed_header_info).
 */
void setup_weed_info(sel_weed_str)
const char *sel_weed_str;
{
    struct weed_header_info *w;
    char *weed_str, *fld, *s;

    /* make a copy we can scribble on safely */
    weed_str = safe_strdup(sel_weed_str);

    w = &Weedlist;
    while ((fld = strtok(weed_str, " \t\n,")) != NULL) {
	weed_str = NULL;

	/* convert "_" to " " */
	for (s = fld ; *s != '\0' ; ++s) {
	    if (*s == '_')
		*s = ' ';
	}

	/* add weeding info to end of list */
	w->next = (struct weed_header_info *)
	    safe_malloc(sizeof(struct weed_header_info));
	w = w->next;
	if (w->hdrskip = (*fld == '!'))
	    ++fld;
	w->hdrname = fld;
	w->hdrlen = strlen(fld);
	w->next = NULL;

	/*
	 * The default weed action is the opposite of the last list entry.
	 * That is, if the list ends in "foo" (an accept action) then the
	 * default is to reject things not in the weed list.  If the list
	 * ends in "!foo" (a reject action) then the default is to accept
	 * things not in the weed list.
	 *
	 * If this doesn't make sense to you, consider the following
	 * two example commands...and what you'd like them to do:
	 *
	 *	readmsg -w "From: Date: Subject:"
	 *	readmsg -w "!From_ !Received: !Status:"
	 *
	 */
	weed_dflt_accept = w->hdrskip;

    }

}


/*
 * Retrieve mailbox location information on a specific message.
 */
int get_mssg_info(minfo_p, mssgno)
struct mssg_info *minfo_p;
int mssgno;
{
    if (mssgno < 1 || mssgno > folder_size) {
	fprintf(stderr, catgets(elm_msg_cat, ReadmsgSet,
	    ReadmsgCannotFindMessage, "%s: Cannot find message number %d.\n"),
	    prog, mssgno);
	return -1;
    }
    minfo_p->folder_pos = folder_idx_list[mssgno-1];
    minfo_p->content_length = 
	(folder_clen_list == NULL ? 0L : folder_clen_list[mssgno-1]);
    return 0;
}

